File: zero-ui-docs/src/js/widget.js
// Customizes widget module pages and widgets in mod lists.

document.addEventListener('DOMContentLoaded', function () {
    editWgtList();
    editWgtSideBar();
    editWgtPage();
});

window.addEventListener('message', function (a) {
    if (a.data.inner_docs !== undefined) {
        onDocsIframeLoaded(a.data.inner_docs);
    }
});

// edit the widget mod item page, changes title and removes the widget tag.
function editWgtPage() {
    let is_mod_pg = false;
    let h1 = null;
    document.querySelectorAll('h1').forEach(function (h) {
        if (h.innerText.trimStart().startsWith("Module ")) {
            h1 = h;
            is_mod_pg = true;
        }
    });

    let is_wgt_pg = false;
    let code = null;
    if (is_mod_pg) {
        document.querySelectorAll('details code').forEach(function (c) {
            if (c.innerText == "widget") {
                code = c;
                is_wgt_pg = true;
            }
        })
    }

    if (is_wgt_pg) {
        // edit page.

        code.remove();
        let txt = h1.childNodes[0].childNodes[0];
        h1.childNodes[0].replaceChild(document.createTextNode("Widget "), txt);

        document.querySelectorAll('h2.location a').forEach(function (a) {
            a.innerText = a.innerText.replace("Module ", "Widget ");
        });

        let inner_docs_a = document.querySelector('a[href="__DOCS/index.html"]');
        let mods_table = inner_docs_a.closest("div.item-table");
        inner_docs_a.closest('div.item-row').remove();
        if (mods_table.querySelector('a') == null) {
            mods_table.previousElementSibling.remove();
            mods_table.remove();
            document.querySelector('a[href="#modules"]').remove();
        }
    }
}

// edit the Modules list of a module, creates a Widgets section, removes widget tag.
function editWgtList() {
    let mods = document.getElementById("modules");
    if (mods == null) {
        return;
    }

    let tags = [];
    mods.nextElementSibling.querySelectorAll("code").forEach(function (c) {
        if (c.innerText == "widget") {
            tags.push(c);
        }
    });

    if (tags.length == 0) {
        return;
    }

    let widgets = document.getElementById("widgets");
    if (widgets == null) {
        widgets = mods.cloneNode(true);
        widgets.id = "widgets";
        let pa = widgets.querySelector("a");
        pa.href = "#widgets";
        pa.innerText = "Widgets";
        let mods_table = mods.nextElementSibling;

        mods.parentElement.insertBefore(widgets, mods_table.nextElementSibling);

        let widgets_table = mods_table.cloneNode(false);
        mods.parentElement.insertBefore(widgets_table, widgets.nextElementSibling);

        // insert sidebar link
        let sidebarMods = document.querySelector("li a[href='#modules']").parentElement;
        let sidebarWgts = sidebarMods.cloneNode(true);
        let sidebarWgtsA = sidebarWgts.querySelector("a");
        sidebarWgtsA.innerText = "Widgets";
        sidebarWgtsA.href = "#widgets";
        sidebarMods.parentElement.insertBefore(sidebarWgts, sidebarMods.nextElementSibling);
    }

    let widgets_table = widgets.nextElementSibling;

    for (const tag of tags) {
        let row = tag.closest("div.item-row");
        tag.remove();
        widgets_table.appendChild(row);
    }

    let mods_table = mods.nextElementSibling;
    if (mods_table.querySelector("a") == null) {
        mods_table.remove();
        mods.remove();
    }
}

// edit the sidebar widget mod items, identified by their tooltip.
function editWgtSideBar() {
    let mods = document.querySelector("div.sidebar-elems div.block.mod");
    if (mods == null) {
        return;
    }

    let wgt_anchors = [];
    mods.querySelector("ul").querySelectorAll("a").forEach(function (a) {
        if (a.title.startsWith("`widget` ")) {
            wgt_anchors.push(a);
        }
    });

    if (wgt_anchors.length == 0) {
        return;
    }

    let widgets = document.createElement("div");
    widgets.classList.add("block");
    widgets.classList.add("mod");

    let widgets_title = mods.querySelector("h3").cloneNode(true);
    let widgets_title_a = widgets_title.querySelector("a");
    widgets_title_a.href = widgets_title_a.href.replace("#modules", "#widgets");
    widgets_title_a.innerText = "Widgets";
    widgets.appendChild(widgets_title);

    let widgets_list = document.createElement("ul");
    for (const prop_anchor of wgt_anchors) {
        prop_anchor.title = prop_anchor.title.replace("`widget` ", "").trimStart();
        widgets_list.appendChild(prop_anchor.parentElement);
    }

    widgets.appendChild(widgets_list);

    mods.parentElement.insertBefore(widgets, mods.nextElementSibling);

    if (mods.querySelector("ul").querySelector("a") == null) {
        mods.remove();
    }
}

function onDocsIframeLoaded(docs) {
    let inner_docs = document.createElement("div");
    inner_docs.innerHTML = docs;
    inner_docs.getElementsByTagName('p')[0].remove();
    inner_docs.getElementsByTagName('p')[0].remove();

    inner_docs.querySelectorAll('a').forEach(function(a) {
        let href = a.getAttribute('href');
        if (href.startsWith('../')) {
            href = href.substring('../'.length);
            a.setAttribute('href', href);
        }
    });

    // convert headers to H5.
    let known_titles = [
        "required-properties",
        "normal-properties",
        "event-properties",
        "state-properties",
        "when-conditions",
    ];
    inner_docs.querySelectorAll('h2,h3,h4').forEach(function (ho) {
        if (!known_titles.includes(ho.id)) {
            let hn = document.createElement('h5');
            hn.innerHTML = ho.innerHTML;

            let id = ho.id;
            if (document.getElementById(id) != null) {
                id += "-inner-docs";
                hn.querySelector("a").href = "#" + id;
            }

            hn.id = id;
            ho.replaceWith(hn);
        }
    });

    // convert property title to H4
    inner_docs.querySelectorAll(".wp-title").forEach(function (s) {
        let place = s.closest('ul');

        s.classList.add("structfield");
        s.classList.add("small-section-header");

        let p_anchor = s.querySelector('a');
        if (p_anchor.href.includes("fn@")) {
            p_anchor.href = "#" + s.id;
        }
        let type_html = "";

        let code = p_anchor.querySelector('code');
        p_anchor.innerText = code.innerText;
        code.innerHTML = p_anchor.outerHTML + type_html;
        p_anchor.remove();
        s.querySelector('strong').appendChild(code);

        let a = document.createElement('a');
        a.classList.add("anchor");
        a.classList.add("field");
        a.href = "#" + s.id;

        s.prepend(a);

        let title = document.createElement("h4");
        while (s.firstChild) {
            title.appendChild(s.firstChild);
        }
        for (index = s.attributes.length - 1; index >= 0; --index) {
            title.attributes.setNamedItem(s.attributes[index].cloneNode());
        }
        title.style.overflowX = "visible";
        title.style.borderBottomWidth = "0";
        title.style.paddingBottom = "0";
        code.style.backgroundColor = "transparent";

        place.replaceWith(title);
    });

    // convert when conditions title to H4
    inner_docs.querySelectorAll("code").forEach(function (c) {
        if (!c.innerText.startsWith("when ")) {
            return;
        }
        let place = c.closest("ul");

        let title_id = c.innerText.replaceAll(/\s+/gm, '-').replaceAll(/[^a-zA-Z0-9_\-\.]/gm, '').toLowerCase();

        c.innerHTML = c.innerHTML.replaceAll(/self\.(\w+)/gm, 'self.<a href="#wp-$1">$1</a>');

        let title = document.createElement("h4");
        title.id = title_id;
        title.classList.add("ww-title");
        title.classList.add("structfield");
        title.classList.add("small-section-header");

        let title_anchor = document.createElement('a');
        title_anchor.classList.add("anchor");
        title_anchor.classList.add("field");
        title_anchor.href = "#" + title_id;

        title.appendChild(title_anchor);
        title.appendChild(c.parentElement);

        title.style.overflowX = "visible";
        title.style.borderBottomWidth = "0";
        title.style.paddingBottom = "0";
        c.style.backgroundColor = "transparent";

        place.replaceWith(title);
    });

    let frame = document.getElementById('wgt-docs-iframe');
    let doc_block = frame.parentElement;
    frame.replaceWith(inner_docs);
    frame.remove();
    while (inner_docs.childNodes.length > 0) {
        doc_block.appendChild(inner_docs.childNodes[0]);
    }

    editWgtPageSideBar();
    if (window.location.hash.length > 0) {
        // scroll to anchor
        window.location.href = window.location.href;
    }

    fetchPropTypes();
}

// customize sidebar of widget module page.
function editWgtPageSideBar() {
    let sidebar = document.querySelector("div.sidebar-elems section");

    let mod_items = sidebar.querySelector('.block');
    let first_mod_item = mod_items.querySelector('a');
    if (first_mod_item != null) {
        let title = document.createElement('h3');
        title.classList.add("sidebar-title");
        let a = first_mod_item.cloneNode(true);
        a.innerText = "Module Items";
        title.appendChild(a);
        mod_items.insertBefore(title, mod_items.querySelector('ul'));
    } else {
        mod_items.remove();
        mod_items = null;
    }

    let widget_items_ul = document.createElement('ul');
    appendSidebarAnchor(widget_items_ul, "required-properties");
    appendSidebarAnchor(widget_items_ul, "normal-properties");
    appendSidebarAnchor(widget_items_ul, "event-properties");
    appendSidebarAnchor(widget_items_ul, "state-properties");
    appendSidebarAnchor(widget_items_ul, "when-conditions");

    let first_widget_item = widget_items_ul.querySelector('a');
    if (first_widget_item != null) {
        let widget_items = document.createElement('div');
        widget_items.classList.add("block");

        let title = document.createElement('h3');
        title.classList.add("sidebar-title");

        let a = first_widget_item.cloneNode(true);
        a.innerText = "Widget Items";
        title.append(a);

        widget_items.appendChild(title);
        widget_items.appendChild(widget_items_ul);
        sidebar.insertBefore(widget_items, mod_items);
    }
}
function appendSidebarAnchor(ul, id) {
    let el = document.getElementById(id);
    if (el != null) {
        let li = document.createElement("li");
        let a = document.createElement("a");
        a.href = "#" + id;
        a.innerText = el.innerText;
        li.appendChild(a);
        ul.appendChild(li);
    }
}

// fetch HTML, parse and fix base URL.
function fetchHtml(url) {
    return fetch(url)
            .then(function(r) { return r.text(); })
            .then(function(html) {
                var parser = new DOMParser();
                var doc = parser.parseFromString(html, "text/html");
                let base = doc.createElement('base');
                base.setAttribute('href', url);
                doc.head.append(base);

                doc.querySelectorAll('a').forEach(function(a) {
                    a.setAttribute('href', a.href);
                });

                return doc;
            });
}

// fetch linked property pages and edit the types with the types.
function fetchPropTypes() {
    let current_page = window.location.href.split('#')[0];
    if (current_page.startsWith("file:///")) {
        return;
    }

    document.querySelectorAll('h4.wp-title').forEach(function (title) {
        let url = title.querySelector('a:not(.anchor)').href;

        if (url.startsWith(current_page)) {
            return;
        }

        url = url.replace('/index.html', '/__DOCS/index.html');
        fetchHtml(url).then(function(doc) {
            if (url.includes('#')) {
                resolvePropPage(title, url, doc);
            } else {
                copyPropType(title, doc);
            }
        });
    });
}
function resolvePropPage(title, url, doc) {
    let anchor = url.split('#')[1];

    let inner_title = doc.getElementById(anchor);
    if (inner_title == null) {
        return;
    }

    let inner_url = inner_title.querySelector('a:not(.anchor)').href.replace('/index.html', '/__DOCS/index.html');

    if (inner_url.includes("fn@")) {
        return;
    }

    fetchHtml(inner_url)
        .then(function(inner_doc) {
            if (inner_url.includes('#')) {
                resolvePropPage(title, inner_url, inner_doc);
            } else {
                copyPropType(title, inner_doc);
            }
        });
}
function copyPropType(title, doc) {
    let target = title.querySelector('code');
    let code = doc.querySelector('pre.rust.fn');
    target.appendChild(code);
    editPropDecl(false, code, code.cloneNode(true));

    let pre = document.createElement('pre');
    pre.classList.add('rust');
    pre.classList.add('fn');
    pre.appendChild(target.firstChild);
    pre.appendChild(document.createTextNode(' =' + code.childNodes[0].textContent.split('=', 2)[1]));
    pre.style.margin = "0";
    pre.style.padding = "0";
    pre.style.backgroundColor = "transparent";
    code.firstChild.remove();
    while (code.firstChild) {
        pre.appendChild(code.firstChild);
    }

    target.replaceWith(pre);

    code.remove();

    let docs_request = title.nextElementSibling.querySelector('span[data-fetch-docs]');
    if (docs_request != null) {
        let docs_target = docs_request.parentElement;
        docs_target.innerHTML = doc.querySelector('summary + .docblock p').innerHTML;
        docs_target.firstChild.remove(); // `property` tag
    }
}

File: zero-ui-docs/src/rs/widget.rs
use std::{
    borrow::Cow,
    cell::{Cell, RefCell},
    collections::{HashMap, HashSet},
    fs,
    path::{Path, PathBuf},
};

use lol_html::html_content::ContentType;
use regex::Regex;

/// Edit Widget module pages and module lists.
pub fn transform(docs_root: &Path) {
    super::util::glob_par_each(docs_root, "**/index.html", |file, mut html| {
        let mut edited = false;
        if let Some(h) = transform_widget_mod_page(&file, &html) {
            html = h;
            edited = true;
        }
        if let Some(h) = transform_mod_list(&html) {
            html = h;
            edited = true;
        }

        if edited {
            fs::write(file, html.as_bytes()).unwrap();
        }
    });
    transform_sidebars(docs_root);
}

fn transform_widget_mod_page(file: &Path, html: &str) -> Option<String> {
    if !html.contains(r#"<div class="docblock"><p><strong><code>widget</code></strong>"#) {
        return None;
    }

    let mut removed_tag = false;
    let modules = Cell::new(false);
    let mut removed_docs_entry = false;
    let mut remove_mods_section = true;
    let mut mod_sidebar_items_count = 0;
    let mut unique_ids = HashSet::new();

    let mut html = super::util::rewrite_html(
        html,
        vec![
            lol_html::text!("h1", |t| {
                if t.as_str() == "Module " {
                    t.replace("Widget ", ContentType::Text);
                }
                Ok(())
            }),
            lol_html::text!("h2", |t| {
                if let Some(m) = t.as_str().strip_prefix("Module ") {
                    let new_t = format!("Widget {m}");
                    t.replace(&new_t, ContentType::Text);
                }
                Ok(())
            }),
            lol_html::element!("div.docblock p strong", |strong| {
                if !removed_tag {
                    strong.remove();
                    removed_tag = true;
                }
                Ok(())
            }),
            lol_html::element!("iframe#wgt-docs-iframe", |iframe| {
                iframe.remove();
                Ok(())
            }),
            lol_html::element!("h2", |h2| {
                if let Some(id) = h2.get_attribute("id") {
                    modules.set(id == "modules");
                }
                Ok(())
            }),
            lol_html::element!("div.item-table div.item-row", |div| {
                if modules.get() {
                    if !removed_docs_entry {
                        removed_docs_entry = true;
                        div.remove();
                    } else {
                        remove_mods_section = false;
                    }
                }
                Ok(())
            }),
            lol_html::element!("div.sidebar-elems li", |_| {
                mod_sidebar_items_count += 1;
                Ok(())
            }),
            lol_html::element!("*", |e| {
                if let Some(id) = e.get_attribute("id") {
                    unique_ids.insert(id);
                }
                Ok(())
            }),
        ],
    )
    .unwrap();

    if remove_mods_section {
        modules.set(false);
        let mut removed_sidebar_entry = false;
        mod_sidebar_items_count -= 1;
        html = super::util::rewrite_html(
            &html,
            vec![
                lol_html::element!("h2", |h2| {
                    if let Some(id) = h2.get_attribute("id") {
                        modules.set(id == "modules");
                        if modules.get() {
                            h2.remove();
                        }
                    }
                    Ok(())
                }),
                lol_html::element!("div.item-table", |div| {
                    if modules.get() {
                        div.remove();
                    }
                    Ok(())
                }),
                lol_html::element!("div.sidebar-elems li", |div| {
                    if !removed_sidebar_entry {
                        removed_sidebar_entry = true;
                        div.remove();
                    }
                    Ok(())
                }),
            ],
        )
        .unwrap();
        unique_ids.remove("modules");
    }

    let docs_file = file.parent().unwrap().join("__DOCS/index.html");
    let docs_html = fs::read_to_string(docs_file).unwrap();

    let first_docblock = Cell::new(true);
    let widget_sidebar_items = RefCell::new(vec![]);
    let mut removed_docblock_help = 0;

    let mut replaced_ids = HashMap::new();
    let known_titles = vec![
        ("required-properties", "Required Properties"),
        ("normal-properties", "Normal Properties"),
        ("event-properties", "Event Properties"),
        ("state-properties", "State Properties"),
        ("when-conditions", "When Conditions"),
    ];

    let docs_html = super::util::rewrite_html(
        &docs_html,
        vec![
            lol_html::element!("*", |e| {
                if let Some(id) = e.get_attribute("id") {
                    if unique_ids.contains(&id) {
                        let new_id = format!("{id}-wp");
                        e.set_attribute("id", &new_id).unwrap();
                        replaced_ids.insert(id, new_id);
                    }
                }
                if let Some(href) = e.get_attribute("href") {
                    if let Some(id) = href.strip_prefix('#') {
                        if let Some(id) = replaced_ids.get(id) {
                            e.set_attribute("href", &format!("#{id}")).unwrap();
                        }
                    }
                }
                Ok(())
            }),
            lol_html::element!("div.docblock", |div| {
                if first_docblock.take() {
                    div.prepend("<!-- COPY ", ContentType::Html);
                    div.append(" -->", ContentType::Html);
                }
                Ok(())
            }),
            lol_html::element!("div.docblock p", |p| {
                if removed_docblock_help < 2 {
                    removed_docblock_help += 1;
                    p.remove();
                }
                Ok(())
            }),
            lol_html::element!("a[href^='../']", |a| {
                let href = a.get_attribute("href").unwrap();
                let href = href.strip_prefix("../").unwrap();

                a.set_attribute("href", href).unwrap();

                Ok(())
            }),
            lol_html::element!("h2", |h2| {
                if let Some(id) = h2.get_attribute("id") {
                    let id = id.strip_suffix("-wp").unwrap_or(&id);
                    if let Some((_, name)) = known_titles.iter().find(|(i, _)| *i == id) {
                        widget_sidebar_items.borrow_mut().push((id.to_owned(), *name));
                    }
                }
                Ok(())
            }),
            lol_html::element!("h1, h2, h3", |h| {
                if let Some(id) = h.get_attribute("id") {
                    if !widget_sidebar_items.borrow().iter().any(|(i, _)| *i == id) {
                        h.set_tag_name("h4").unwrap();
                    }
                }
                Ok(())
            }),
        ],
    )
    .unwrap();

    let inner_html = Regex::new(r#"<!-- COPY \s*((?s).+?)\s* -->"#).unwrap();
    let inner_html = &inner_html.captures(&docs_html).unwrap()[1];

    let mut summaries = HashMap::new();

    // matches:
    // <ul>\n<li><span id='{id}' class='wp-title'><strong><a href="{href}"><code>{name}</code></a></strong></span></li>\n</ul>
    let property_titles = Regex::new(r##"(?s)<ul>\s*<li><span id='(?P<id>[a-z\-_]+)' class='wp-title'><strong><a href="(?P<href>[\./a-z#\-_@]+)"><code>(?P<name>[a-z_]+)</code></a></strong></span></li>\s*</ul>"##).unwrap();
    let inner_html = property_titles.replace_all(inner_html, |caps: &regex::Captures| {
        let id = &caps["id"];
        let href = &caps["href"];
        let name = &caps["name"];

        let mut prop_types = String::new();

        if let Some(pfn_file) = resolve_property_href(file, href) {
            if let Ok(pfn_html) = fs::read_to_string(&pfn_file) {
                let base_path = file.parent().unwrap();
                let pfn_base_path = pfn_file.parent().unwrap();
                let found_summary = Cell::new(false);
                let pfn_html = super::util::rewrite_html(&pfn_html, vec![
                    lol_html::element!("pre.rust.fn code", |code| {
                        code.prepend("<!-- CODE ", ContentType::Html);
                        code.append(" -->", ContentType::Html);
                        Ok(())
                    }),
                    lol_html::element!("a", |a| {
                        if let Some(href) = a.get_attribute("href") {
                            if !href.contains("://") && !href.starts_with("//") {
                                let full = pfn_base_path.join(href);
                                if let Some(href) = pathdiff::diff_paths(full, base_path) {
                                    a.set_attribute("href", href.display().to_string().as_ref()).unwrap();
                                }
                            }
                        }
                        Ok(())
                    }),
                    // summary + .docblock p
                    lol_html::element!("summary", |_| {
                        found_summary.set(true);
                        Ok(())
                    }),
                    lol_html::element!(".docblock p", |code| {
                        if found_summary.take() {
                            code.prepend("<!-- SUMMARY ", ContentType::Html);
                            code.append(" -->", ContentType::Html);
                        }                        
                        Ok(())
                    })
                ]).unwrap();

                let code = Regex::new(r"<!-- CODE \s*((?s).+?)\s* -->").unwrap();
                let code_inner_html = &code.captures(&pfn_html).unwrap()[1];

                if let Some((before_eq, after_eq)) = code_inner_html.split_once('=') {
                    let valid_prop = Regex::new(r#"^\s*(pub(\(\w+\))?)?\s+\w+\s*$"#).unwrap();
                    if valid_prop.is_match(before_eq) {
                        prop_types.push_str(" = ");
                        prop_types.push_str(after_eq.trim_start());
                    }
                }

                let summary = Regex::new(r"<!-- SUMMARY \s*((?s).+?)\s* -->").unwrap();
                if let Some(summary) = summary.captures(&pfn_html).map(|c| c[1].to_owned()) {
                    summaries.insert(name.to_owned(), summary);
                }
            }
        }

        let mut href = Cow::Borrowed(href);
        if href.contains("fn@") {
            href = Cow::Owned(format!("#{id}"));
        }

        format!(r##"<h3 id="{id}" class="wp-title variant small-section-header" style="overflow-x:visible;"><a href="#{id}" class="anchor field"></a><code style="background-color:transparent;"><a href="{href}">{name}</a>{prop_types}</code></h3>"##)
    });
    let inner_html = inner_html.as_ref();

    // matches:
    // <span data-fetch-docs='{name}'></span>
    let summary_fetch = Regex::new(r##"<span data-fetch-docs='(?P<name>[a-z_]+)'></span>"##).unwrap();
    let inner_html = summary_fetch.replace_all(inner_html, |caps: &regex::Captures| {
        let name = &caps["name"];

        summaries.remove(name).unwrap_or_default()
    });
    let inner_html = inner_html.as_ref();

    // matches:
    // <ul>\n<li><strong><code>when {expr}</code></strong></li>\n</ul>
    let when_titles = Regex::new(r##"(?s)<ul>\s*<li><strong><code>(?P<expr>.+?)</code></strong></li>\s*</ul>"##).unwrap();
    let spaces = Regex::new(r"(?s)\s+").unwrap();
    let when_id_invalids = Regex::new(r"[^a-zA-Z0-9_\-\.]").unwrap();
    let when_self_prop = Regex::new(r"self\.(?P<name>[\w_]+)").unwrap();
    let inner_html = when_titles.replace_all(inner_html, |caps: &regex::Captures| {
        let expr = &caps["expr"];
        let id = spaces.replace_all(expr, "-");
        let id = when_id_invalids.replace_all(&id,  "");

        let expr = when_self_prop.replace_all(expr, |caps: &regex::Captures| {
            let name = &caps["name"];
            format!(r##"self.<a href="#wp-{name}">{name}</a>"##)
        });

        format!(r##"<h3 id="{id}" class="wp-title variant small-section-header" style="overflow-x:visible;"><a href="#{id}" class="anchor field"></a><code style="background-color:transparent;">{expr}</code></h3>"##)
    });
    let inner_html = inner_html.as_ref();

    let mut sidebar_add = String::new();
    let widget_sidebar_items = widget_sidebar_items.into_inner();
    if !widget_sidebar_items.is_empty() {
        sidebar_add.push_str(r#"<div class="block"><h3 class="sidebar-title">Widget Items</h3><ul>"#);
        for (id, label) in widget_sidebar_items {
            sidebar_add.push_str(r##"<li><a href="#"##);
            sidebar_add.push_str(&id);
            sidebar_add.push_str(r#"">"#);
            sidebar_add.push_str(label);
            sidebar_add.push_str("</a></li>");
        }
        sidebar_add.push_str("</ul></div>");
    }
    if mod_sidebar_items_count > 0 {
        sidebar_add.push_str(r#"<h3 class="sidebar-title">Module Items</h3>"#);
    }

    first_docblock.set(true);
    html = super::util::rewrite_html(
        &html,
        vec![
            lol_html::element!("div.docblock", move |div| {
                if first_docblock.take() {
                    div.append(inner_html, ContentType::Html);
                }
                Ok(())
            }),
            lol_html::element!("div.sidebar-elems section", |section| {
                section.prepend(&sidebar_add, ContentType::Html);
                Ok(())
            }),
        ],
    )
    .unwrap();

    Some(html)
}
fn resolve_property_href(mod_file: &Path, href: &str) -> Option<PathBuf> {
    let mod_dir = mod_file.parent().unwrap();
    if let Some(name) = href.strip_prefix("fn@") {
        let pfn_file = format!("__DOCS/fn.__p_{name}.html");
        let cap_only_file = mod_dir.join(pfn_file);
        Some(cap_only_file)
    } else if let Some((inner_mod_file, id)) = href.rsplit_once('#') {
        let inner_mod_file = mod_dir.join(inner_mod_file);
        let inner_mod_dir = inner_mod_file.parent().unwrap();
        // use __DOCS to avoid reading from another widget being modified.
        let file = inner_mod_dir.join("__DOCS/index.html");
        if let Ok(html) = fs::read_to_string(&file) {
            let mut href = None;
            super::util::analyze_html(
                &html,
                vec![lol_html::element!(format!("span#{id} a"), |a| {
                    href = a.get_attribute("href");
                    Ok(())
                })],
            );

            href.and_then(|href| {
                let href = href.strip_prefix("../").unwrap_or(&href);
                resolve_property_href(inner_mod_file.as_ref(), href)
            })
        } else {
            None
        }
    } else if href.contains("/fn.") {
        Some(mod_dir.join(href))
    } else {
        None
    }
}

fn transform_mod_list(html: &str) -> Option<String> {
    let modules = Cell::new(false);
    let mod_entry = Cell::new(0);
    let mut tagged_mods = HashSet::new();

    super::util::analyze_html(
        html,
        vec![
            lol_html::element!("h2", |h2| {
                if let Some(id) = h2.get_attribute("id") {
                    modules.set(id == "modules");
                }
                Ok(())
            }),
            lol_html::element!("div.item-row", |_| {
                if modules.get() {
                    mod_entry.set(mod_entry.get() + 1);
                }
                Ok(())
            }),
            lol_html::text!("div.item-row code", |t| {
                if modules.get() && t.as_str() == "widget" {
                    tagged_mods.insert(mod_entry.get());
                }
                Ok(())
            }),
        ],
    );

    if tagged_mods.is_empty() {
        return None;
    }

    let move_all_mods = tagged_mods.len() == mod_entry.get();

    modules.set(false);
    let rmv_strong = Cell::new(false);

    let mut transforms = if move_all_mods {
        vec![
            lol_html::element!("h2", |h2| {
                if let Some(id) = h2.get_attribute("id") {
                    modules.set(id == "modules");
                    if modules.get() {
                        h2.set_attribute("id", "widgets").unwrap();
                    }
                }
                Ok(())
            }),
            lol_html::element!("a[href='#modules']", |a| {
                a.set_attribute("href", "#widgets").unwrap();
                Ok(())
            }),
            lol_html::text!("a[href='#widgets']", |t| {
                if t.as_str() == "Modules" {
                    t.replace("Widgets", ContentType::Text);
                }
                Ok(())
            }),
            lol_html::element!("div.item-row", |_| {
                if modules.get() {
                    rmv_strong.set(true);
                }
                Ok(())
            }),
        ]
    } else {
        mod_entry.set(0);

        vec![
            lol_html::element!("h2", |h2| {
                if let Some(id) = h2.get_attribute("id") {
                    modules.set(id == "modules");
                }
                Ok(())
            }),
            lol_html::element!("div.item-row", |div| {
                if modules.get() {
                    mod_entry.set(mod_entry.get() + 1);
                    if tagged_mods.contains(&mod_entry.get()) {
                        rmv_strong.set(true);
                        div.before("<!-- CUT ", ContentType::Html);
                        div.after(" -->", ContentType::Html);
                    }
                }
                Ok(())
            }),
        ]
    };
    transforms.push(lol_html::element!("div.item-row strong", |s| {
        if rmv_strong.take() {
            s.remove();
        }
        Ok(())
    }));

    let mut html = super::util::rewrite_html(html, transforms).unwrap();

    if !move_all_mods {
        let cut = Regex::new(r"<!-- CUT \s*((?s).+?)\s* -->").unwrap();

        let mut widgets =
            r##"<h2 id="widgets" class="small-section-header"><a href="#widgets">Widgets</a></h2><div class="item-table">"##.to_owned();
        for cap in cut.captures_iter(&html) {
            widgets.push_str(&cap[1]);
        }
        widgets.push_str("</div>");

        modules.set(false);

        html = super::util::rewrite_html(
            &html,
            vec![
                lol_html::element!("h2", |h2| {
                    if let Some(id) = h2.get_attribute("id") {
                        modules.set(id == "modules");
                    }
                    Ok(())
                }),
                lol_html::element!("div.item-table", |div| {
                    if modules.get() {
                        div.after(&widgets, ContentType::Html);
                    }
                    Ok(())
                }),
                lol_html::comments!("div.item-table", |c| {
                    if c.text().contains("CUT") {
                        c.remove();
                    }
                    Ok(())
                }),
                lol_html::element!(".sidebar-elems a[href='#modules']", |a| {
                    let props = r##"</li><li><a href="#widgets">Widgets</a>"##;
                    a.after(props, ContentType::Html);
                    Ok(())
                }),
            ],
        )
        .unwrap();
    }

    Some(html)
}

/// Edit sidebar lists, creates a new "Widgets" section.
fn transform_sidebars(docs_root: &Path) {
    super::util::glob_par_each(docs_root, "**/sidebar-items.js", |file, js| {
        if !js.starts_with("initSidebarItems(") {
            return;
        }

        let edit = js.contains("`widget` ");

        if edit {
            // TODO

            fs::write(file, js.as_bytes()).unwrap();
        }
    });
}
